scipy/signal/signaltools.py

```diff
@@ -777,6 +777,88 @@ def convolve(in1, in2, mode='full', method='auto'):
     return correlate(volume, _reverse_and_conj(kernel), mode, 'direct')
 
 
+def circular_convolve(in1, in2, period):
+    """
+    Circular convolution of two 1-dimensional arrays.
+
+    Circular convolve `in1` and `in2` with given `period`.
+
+    Parameters
+    ----------
+    in1 : array_like, 1-D
+        First input.
+    in2 : array_like, 1-D
+        Second input. Should have the same number of dimensions as `in1`.
+    period : int
+        Period of the circular convolution.
+
+    Returns
+    -------
+    result : array, 1-D
+        A 1-dimensional array containing the result of the circular
+        convolution of `in1` with `in2`.
+
+    See Also
+    --------
+    convolve
+
+    Notes
+    -----
+    The (modulo-M) circular/cyclic/periodic convolution of period :math:`M`
+    of the two signals :math:`x[k]` and :math:`h[k]` is defined as
+
+    .. math::
+
+        y[k] = \sum_{\kappa=0}^{M-1} \tilde{x}[k - \kappa] \; \tilde{h}[\kappa]
+
+    where the periodic summations :math:`\tilde{x}[k]` and `\tilde{h}[\kappa]`
+    of :math:`x[k]` and :math:`x[k]` are defined as
+
+    .. math::
+
+        \tilde{x}[k] &= \sum_{m = -\infty}^{\infty} x[m \cdot M + k] \\
+        \tilde{h}[k] &= \sum_{m = -\infty}^{\infty} h[m \cdot M + k]
+
+    Examples
+    --------
+    Equivalence of circular and linear convolution:
+
+    >>> from scipy import signal
+    >>> a = np.ones(5)
+    >>> b = np.ones(5)
+    >>> circular_convolve(a, b, 5)
+    array([ 5.,  5.,  5.,  5.,  5.])
+    >>> np.convolve(a, b, mode='full')
+    array([ 1.,  2.,  3.,  4.,  5.,  4.,  3.,  2.,  1.])
+    >>> circular_convolve(a, b, 9)
+    array([ 1.,  2.,  3.,  4.,  5.,  4.,  3.,  2.,  1.])
+    """
+    in1 = _periodic_summation(in1, period)
+    in2 = _periodic_summation(in2, period)
+
+    return np.fromiter([np.dot(np.roll(in1[::-1], k+1), in2)
+                        for k in np.arange(period)], float)
```

Comment:
I suspect this will be quite slow for long signals. Have you considered using something like `ifft(fft(in1, period) * fft(in2, period))`?

Reply:
Yes the numerical complexity will be high for long signals. However I wanted to stay in line with the usual convention used for the familiy of convolve functions (e.g. `numpy.convolve`, `conv` in MATLAB, ...) which implement the convolution by summation.